lm_human_preference_details/train_reward_jax.py

```diff
@@ -0,0 +1,765 @@
+            if self.start_token is not None:
+                try:
+                    first_index = tokens.index(self.start_token) + 1
+                    if first_index < len(tokens):
+                        tokens = tokens[first_index:]
+                except:
+                    continue
+            tokens = tokens[: self.query_length]
+            if self.end_token is not None:
+                try:
+                    last_index = len(tokens) - tokens[::-1].index(self.end_token)
+                    tokens = tokens[:last_index]
+                except:
+                    continue
+            output = self.tokenizer.pad(
+                {"input_ids": tokens},
+                padding="max_length",
+                max_length=self.query_length,
+                return_tensors="pt",
+                return_attention_mask=True,
+            )
+            yield output
+
+
+def right_padding_to_left_padding(tokens, pad_id):
+    """Convert from right padding to left padding."""
+    assert tokens.ndim == 2
+    return np.array(
+        [
+            [pad_id] * (row == pad_id).sum() + [x for x in row if x != pad_id]
+            for row in tokens
+        ]
+    )
+
+
+def ceil_div(a, b):
+    return (a - 1) // b + 1
+
+
+def exact_div(a, b):
+    q = a // b
+    if a != q * b:
+        raise ValueError(f"Inexact division: {a} / {b} = {a/b}")
+    return q
+
+
+# TODO: pmap `generate` to accelerate reward model normalization?
+def generate(pretrained_model, queries, args, generation_config):
+    """generate in a way that does not affect padding tokens"""
+    context_length = queries.shape[1]
+    attention_mask = queries != args.pad_token_id
+    # set padding tokens to 0
+    input_ids = jnp.where(attention_mask, queries, 0)
+    output = pretrained_model.generate(
+        input_ids=input_ids,
+        attention_mask=attention_mask.astype("int32"),
+        # position_ids=attention_mask.cumsum(1) - attention_mask.long(),
+        # generation collapsed if this was turned on.
+        # TODO: why does generation collapse with this?
+        generation_config=generation_config,
+        return_dict_in_generate=True,
+    )
+    return jnp.concatenate((queries, output.sequences[:, context_length:]), axis=1)
+
+
+def single_epoch_linear_schedule(global_step, args):
+    """ anneal learning rate linearly to reach 0 after one epoch."""
+    frac = 1.0 - global_step * args.batch_size / args.labels.num_train
+    return args.lr * frac
+
+
+def create_initial_reward_state_and_models(init_key, args):
+    # pylint: disable=redefined-outer-name
+    """reate reward model and initial reward state."""
+
+    reward_backbone = transformers.FlaxAutoModelForCausalLM.from_pretrained(
+        args.base_model
+    )
+    reward_head = RewardHead(head_input_size=reward_backbone.config.hidden_size)
+
+    if args.use_tensorflow_adam:
+        raise NotImplementedError("tensorflow adam is not implemented yet.")
```

Comment:
Btw optax's adam can be made equivalent to tensorflow's adam with the following optimizer. We should include this `use_tensorflow_adam` in the jax version as well.

https://gist.github.com/vwxyzjn/7005a81ba39deb3bc8043041bd715be1#file-main_jax-py-L69-L78